(IO(Http) ? HttpRequest) does not reply if the URL is malformed

Due to a configuration error, my Spray app made an http request to a malformed URL.

Internally to Spray, an exception was thrown, but Spray did not reply with a Failure, instead the exception was swallowed and my Ask timed out.

Repro:

```
import org.specs2.mutable.SpecificationLike
import spray.http.Uri
import akka.testkit.TestKit
import spray.can.Http
import akka.io.IO
import spray.httpx.RequestBuilding
import akka.pattern.{AskTimeoutException, ask}
import scala.concurrent.Await
import org.specs2.time.NoTimeConversions
import scala.concurrent.duration._
import spray.http.Uri.Host

class AkkaIoHttpSpec
  extends TestKit(ActorSystem())
  with SpecificationLike
  with RequestBuilding
  with NoTimeConversions {

  "IO(Http)" should {
    "always reply with success or failure, even if the URL is malformed" in {

      val badUrl = Uri("example.com:8080/myPath")

      badUrl.authority.host mustEqual Host.Empty
      badUrl.authority.port mustEqual 0
      badUrl.scheme mustEqual "example.com"

      badUrl.toString() mustEqual "example.com:8080/myPath"

      val io = IO(Http)

      val response = (io ? Get(badUrl))(1.second)

      // BUG: this shouldn't time out, but ought to hand us the underlying
      // spray.http.IllegalUriException
      Await.result(response, 2.second) must throwAn[AskTimeoutException]
    }
  }
}
```

Logs for the above test:

```
2014-06-02 13:01:27,035  INFO  akka.event.slf4j.Slf4jLogger  - Slf4jLogger started
2014-06-02 13:01:27,049 12:01:27.044UTC DEBUG akka.event.EventStream EventStream(akka://AkkaIoHttpSpec) - logger log1-Slf4jLogger started
2014-06-02 13:01:27,055 12:01:27.046UTC DEBUG akka.event.EventStream EventStream(akka://AkkaIoHttpSpec) - Default Loggers started
2014-06-02 13:01:28,261 12:01:28.241UTC ERROR akka.actor.OneForOneStrategy akka://AkkaIoHttpSpec/user/IO-HTTP/host-connector-0 - Invalid port -1
akka.actor.ActorInitializationException: exception during creation
        at akka.actor.ActorInitializationException$.apply(Actor.scala:164) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.actor.ActorCell.create(ActorCell.scala:596) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:456) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.actor.ActorCell.systemInvoke(ActorCell.scala:478) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:263) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.dispatch.Mailbox.run(Mailbox.scala:219) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393) [akka-actor_2.11-2.3.3.jar:na]
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260) [scala-library-2.11.0.jar:na]
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339) [scala-library-2.11.0.jar:na]
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979) [scala-library-2.11.0.jar:na]
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107) [scala-library-2.11.0.jar:na]
Caused by: spray.http.IllegalUriException: Invalid port -1
        at spray.http.Uri$.fail(Uri.scala:765) ~[spray-http_2.11-1.3.1-20140423.jar:na]
        at spray.http.Uri$.normalizePort(Uri.scala:726) ~[spray-http_2.11-1.3.1-20140423.jar:na]
        at spray.http.Uri$Authority.normalizedFor(Uri.scala:349) ~[spray-http_2.11-1.3.1-20140423.jar:na]
        at spray.http.Uri$Authority.normalizedForHttp(Uri.scala:347) ~[spray-http_2.11-1.3.1-20140423.jar:na]
        at spray.can.client.HttpHostConnector.<init>(HttpHostConnector.scala:40) ~[spray-can_2.11-1.3.1-20140423.jar:na]
        at spray.can.HttpManager$$anonfun$4.apply(HttpManager.scala:199) ~[spray-can_2.11-1.3.1-20140423.jar:na]
        at spray.can.HttpManager$$anonfun$4.apply(HttpManager.scala:199) ~[spray-can_2.11-1.3.1-20140423.jar:na]
        at akka.actor.TypedCreatorFunctionConsumer.produce(Props.scala:340) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.actor.Props.newActor(Props.scala:249) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.actor.ActorCell.newActor(ActorCell.scala:552) ~[akka-actor_2.11-2.3.3.jar:na]
        at akka.actor.ActorCell.create(ActorCell.scala:578) ~[akka-actor_2.11-2.3.3.jar:na]
        ... 9 common frames omitted
2014-06-02 13:01:28,262 12:01:28.254UTC INFO  akka.actor.LocalActorRef akka://AkkaIoHttpSpec/user/IO-HTTP/host-connector-0 - Message [spray.http.HttpRequest] from Actor[akka://AkkaIoHttpSpec/temp/$a] to Actor[akka://AkkaIoHttpSpec/user/IO-HTTP/host-connector-0#-109647337] was not delivered. [1] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
2014-06-02 13:01:28,263 12:01:28.258UTC DEBUG spray.can.HttpManager akka://AkkaIoHttpSpec/user/IO-HTTP - received AutoReceiveMessage Envelope(Terminated(Actor[akka://AkkaIoHttpSpec/user/IO-HTTP/host-connector-0#-109647337]),Actor[akka://AkkaIoHttpSpec/user/IO-HTTP/host-connector-0#-109647337])
```

I think that Spray ought to reply to the Ask with the underlying `spray.http.IllegalUriException`, rather than timing out.

--- a/spray-can-tests/src/test/scala/spray/can/client/SprayCanClientSpec.scala
+++ b/spray-can-tests/src/test/scala/spray/can/client/SprayCanClientSpec.scala
@@ -440,6 +440,17 @@ class SprayCanClientSpec extends Specification with NoTimeConversions {
       probe.expectMsgType[Status.Failure].cause.getMessage must startWith("Cannot establish effective request URI")
     }
 
+    "produce an error if the request is missing a scheme" in {
+      val probe = TestProbe()
+      probe.send(IO(Http), Get("example.com:8080/abc"))
+      probe.expectMsgType[Status.Failure].cause.getMessage must startWith("requirement failed: Not a valid HTTP URI scheme: 'example.com'")
+    }
+    "produce an error if the request URI scheme isn't 'http' or 'https' one" in {
+      val probe = TestProbe()
+      probe.send(IO(Http), Get("blub:example.com:8080/abc"))
+      probe.expectMsgType[Status.Failure].cause.getMessage must startWith("requirement failed: Not a valid HTTP URI scheme: 'blub'")
+    }
+
     "produce an error if the request was not completed within the configured timeout" in new TestSetup {
       val probe = TestProbe()
       probe.send(IO(Http), Get("/abc") ~> Host(hostname, port))

--- a/spray-can/src/main/scala/spray/can/HttpManager.scala
+++ b/spray-can/src/main/scala/spray/can/HttpManager.scala
@@ -164,6 +164,8 @@ private[can] class HttpManager(httpSettings: HttpExt#Settings) extends Actor wit
     }
 
   def connectorForUri(uri: Uri) = {
+    require(uri.scheme == "http" || uri.scheme == "https",
+      s"Not a valid HTTP URI scheme: '${uri.scheme}' in '$uri'. (Did you forget http:// ?)")
     val host = uri.authority.host
     connectorFor(HostConnectorSetup(host.toString, uri.effectivePort, sslEncryption = uri.scheme == "https"))
   }

